zkevm-circuits/src/evm_circuit/execution/error_oog_constant.rs

```diff
@@ -0,0 +1,212 @@
+        )
+    }
+
+    fn test_oog_constant(tx: eth_types::Transaction, is_success: bool) {
+        let code = if is_success {
+            bytecode! {
+                PUSH1(0)
+                PUSH1(0)
+                RETURN
+            }
+        } else {
+            bytecode! {
+                PUSH1(0)
+                PUSH1(0)
+            }
+        };
+
+        // Get the execution steps from the external tracer
+        let block: GethData = TestContext::<2, 1>::new(
+            None,
+            account_0_code_account_1_no_code(code),
+            |mut txs, _accs| {
+                txs[0]
+                    .to(tx.to.unwrap())
+                    .from(tx.from)
+                    .gas_price(tx.gas_price.unwrap())
+                    .gas(tx.gas - 1)
+                    .input(tx.input)
+                    .value(tx.value);
+            },
+            |block, _tx| block.number(0xcafeu64),
+        )
+        .unwrap()
+        .into();
+
+        let mut builder = BlockData::new_from_geth_data(block.clone()).new_circuit_input_builder();
+
+        builder
+            .handle_block(&block.eth_block, &block.geth_traces)
+            .unwrap();
+        let block = block_convert(&builder.block, &builder.code_db);
+        assert_eq!(run_test_circuit_incomplete_fixed_table(block), Ok(()));
+    }
+
+    fn mock_tx(value: Word, gas_price: Word, calldata: Vec<u8>) -> eth_types::Transaction {
+        let from = MOCK_ACCOUNTS[1];
+        let to = MOCK_ACCOUNTS[0];
+        eth_types::Transaction {
+            from,
+            to: Some(to),
+            value,
+            gas: gas(&calldata),
+            gas_price: Some(gas_price),
+            input: calldata.into(),
+            ..Default::default()
+        }
+    }
+
+    #[test]
+    fn oog_constant_gadget_simple() {
+        // Transfer 1 ether, successfully
+        // in root call
+        test_oog_constant(mock_tx(eth(1), gwei(2), vec![]), false);
+        // TODO: add internal call test
+    }
+
+    #[test]
+    fn stack_overflow_simple() {
+        test_stack_overflow(OpcodeId::PUSH1, &[1]);
+    }
+
+    fn test_stack_overflow(opcode: OpcodeId, bytes: &[u8]) {
+        assert!(bytes.len() as u8 == opcode.as_u8() - OpcodeId::PUSH1.as_u8() + 1,);
+
+        let mut bytecode = bytecode! {
+            .write_op(opcode)
+        };
+        for b in bytes {
+            bytecode.write(*b, false);
+        }
+        // still add 1024 causes stack overflow
+        for _ in 0..1025 {
+            bytecode.write_op(opcode);
+            for b in bytes {
+                bytecode.write(*b, false);
+            }
+        }
+        // append final stop op code
+        bytecode.write_op(OpcodeId::STOP);
+
+        assert_eq!(
+            run_test_circuits(
+                TestContext::<2, 1>::simple_ctx_with_bytecode(bytecode).unwrap(),
+                None
+            ),
+            Ok(())
+        );
+    }
+
+    fn test_stack_underflow(value: Word) {
```

Comment:
what about moving these to push/pop gadget file?

Reply:
My original plan is moving to real implementing stack overflow/underflow gadget file in the future. but OK to me for now

Reply:
moved in [1f219ec](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/647/commits/1f219ec84675ffe570ac3da396adb5140f0e0e0d)